### PR TITLE
POEM_086 Implementation: model_options

### LINKED ISSUE #2930: POEM 086 Implementation: Problem.model_options
### Desired capability or behavior.

Implements POEM_086, which adds `Problem.model_options`, allowing options to be passed down through the model tree to whichever components can accept them.  Target systems for options are specified through a glob pattern.

### Associated POEM

086

### PR DESCRIPTION
**Pull request by robfalck**
### Summary

Adds Problem.model_options, allowing options to be passed down to any system in the model hierarchy.

### Related Issues

- Resolves #2930 

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: openmdao/docs/openmdao_book/features/core_features/options/options.ipynb [hunk 1]
```diff
@@ -503,12 +503,212 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setting Multiple Option Values Simultaneously\n",
+    "\n",
+    "The *OptionsDictionary* class is versatile and can be used in a variety of situations.\n",
+    "While all systems contain an `options` attribute, users may wish to introduce other options for their systems or other classes using OpenMDAO's *OptionsDictionary*.\n",
+    "\n",
+    "After initialization, users might want the ability to modify more than one options at a time.\n",
```
Comment (naylor-b) [inline openmdao/docs/openmdao_book/features/core_features/options/options.ipynb:516]: 'more than one options'  -->  'more than one option'

**Review by naylor-b — APPROVED** [review]

File: openmdao/docs/openmdao_book/features/core_features/options/options.ipynb [hunk 2]
```diff
@@ -503,12 +503,212 @@
...
+   "metadata": {},
+   "source": [
+    "## Setting options throughout a problem model (`Problem.model_options`)\n",
+    "\n",
+    "It is common for options to be passed down through a model tree from parent Group to child Group or Component.\n",
+    "This can make the addition of a new option deep in the model a tedious procedure, since each parent system may need to have an option added to itself in order to know how to instantiate its children, and then actually take the given input option and pass it to the child.\n",
+    "\n",
+    "Different options are available to alleviate this. For defining the options, it might be useful to have a common function that is referenced by all pertinent classes which adds the necessary options. Alternatively, those classes can subclass from a class which automatically adds the appropriate options.\n",
+    "\n",
+    "For passing options to components nested deeply within the model, the OpenMDAO `Problem` object provides an attribute named `model_options`. `Problem.model_options` is a standard dictionary, keyed by a string that serves as a glob filter for system pathnames.\n",
```
Comment (swryan) [inline openmdao/docs/openmdao_book/features/core_features/options/options.ipynb:567]: suggest hyperlinking "glob filter" to a definition, e.g. https://en.wikipedia.org/wiki/Glob_(programming)

File: openmdao/core/problem.py [hunk 3]
```diff
@@ -171,6 +171,12 @@ class Problem(object):
         publishing work that uses this class.
     options : <OptionsDictionary>
         Dictionary with general options for the problem.
+    model_options : dict
+        A  dictionary of options to be passed to subsystems in the problem's model.
+        This dictionary is keyed by a path pattern string, and the associated value for each path
+        pattern is a dictionary of {option_name: option_val}. Those subsystems within the
+        hierarchy which match the path pattern and that have an option of the given name, will
+        have the value of that option overridden by value given in the dictionary.
```
Comment (swryan) [inline openmdao/core/problem.py:179]: should probably clarify that this happens at setup time..  setting an option after setup will have no effect...

File: openmdao/core/group.py [hunk 4]
```diff
@@ -4345,3 +4345,18 @@ def _ordered_comp_name_iter(self):
                 yield from s._ordered_comp_name_iter()
             else:
                 yield s.pathname
+
+    @property
+    def model_options(self):
```
Comment (swryan) [inline openmdao/core/group.py:4350]: I don't see that this property on `Group` is ever used.. is there a use case for it? is it in need of a unit test?

File: openmdao/docs/openmdao_book/features/core_features/options/options.ipynb [hunk 5]
```diff
@@ -562,9 +562,11 @@
     "It is common for options to be passed down through a model tree from parent Group to child Group or Component.\n",
     "This can make the addition of a new option deep in the model a tedious procedure, since each parent system may need to have an option added to itself in order to know how to instantiate its children, and then actually take the given input option and pass it to the child.\n",
     "\n",
-    "Different options are available to alleviate this. For defining the options, it might be useful to have a common function that is referenced by all pertinent classes which adds the necessary options. Alternatively, those classes can subclass from a class which automatically adds the appropriate options.\n",
+    "Different options are available to address this. \n",
```
Comment (swryan) [inline openmdao/docs/openmdao_book/features/core_features/options/options.ipynb:565]: this sentence is now redundant?

**Review by swryan — APPROVED** [review]

File: openmdao/docs/openmdao_book/features/core_features/options/options.ipynb [hunk 6]
```diff
@@ -22,11 +22,11 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Component Options (Arguments to Components)\n",
+    "# Systems Options (Arguments to Components and Groups)\n",
```
Comment (swryan) [inline openmdao/docs/openmdao_book/features/core_features/options/options.ipynb:25]: I think this should be `System` (singular)

**Pull request merged**